### PR TITLE
Add pause and resume methods to the core.tcpsocket implementation

### PR DESCRIPTION
**Pull request by bemasc**
Passes the freedom integration test.

### CODE REVIEW

File: providers/client_socket.js [hunk 1]
```diff
@@ -122,6 +127,15 @@ ClientSocket.prototype.write = function(data) {
   this.outputStream.write(stringData, stringData.length);
 };
 
+ClientSocket.prototype.pause = function() {
+  this.paused = true;
+};
+
+ClientSocket.prototype.resume = function() {
+  this.paused = false;
+  this.rawInputStream.asyncWait(this.inputStreamCallback, 0, 0, mainThread);
```
Comment (willscott) [inline providers/client_socket.js]: seems plausible. are there flags we want to the input stream https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsITransport#openInputStream() that would make the back pressure more effective (by e.g. restricting how much can be in queue, which firefox maybe uses to define its tcp window size?)

File: providers/core.tcpsocket.js [hunk 2]
```diff
@@ -75,6 +75,16 @@ Socket_firefox.prototype.write = function(buffer, continuation) {
   }
 };
 
+Socket_firefox.prototype.pause = function(continuation) {
+  this.clientSocket.pause();
```
Comment (willscott) [inline providers/core.tcpsocket.js]: this would fail if called on a server socket, right? wrap in a catch or error early?

**Comment by willscott** [comment]
Looks good! couple minor nits.

File: providers/client_socket.js (hunk 1 above)
Comment (bemasc) [inline providers/client_socket.js]: I don't know what to do with any of those flags.  I expect the defaults are fine.  If you want I can add a TODO to investigate...

File: providers/core.tcpsocket.js (hunk 2 above)
Comment (bemasc) [inline providers/core.tcpsocket.js]: Done.

**Pull request merged**